planner/core/find_best_task.go

```diff
@@ -196,6 +197,135 @@ func (ds *DataSource) tryToGetDualTask() (task, error) {
 	return nil, nil
 }
 
+// candidatePath is used to maintain required info for skyline pruning.
+type candidatePath struct {
+	path         *accessPath
+	columnSet    *intsets.Sparse // columnSet is the set of columns that occurred in the access conditions.
+	isSingleScan bool
+	isMatchProp  bool
+}
+
+// compareColumnSet will compares the two set. The last return value is used to indicate
+// if they are comparable, it is false when both two sets have columns that do not occur in the other.
+// When the second return value is true, the value of first:
+// (1) -1 means that `l` is a strict subset of `r`;
+// (2) 0 means that `l` equals to `r`;
+// (3) 1 means that `l` is a strict superset of `r`.
+func compareColumnSet(l, r *intsets.Sparse) (int, bool) {
+	lLen, rLen := l.Len(), r.Len()
+	if lLen <= rLen {
+		if isSubset := l.SubsetOf(r); !isSubset {
+			return 0, false
+		}
+		if lLen == rLen {
+			return 0, true
+		}
+		return -1, true
+	}
+	if isSubset := r.SubsetOf(l); !isSubset {
+		return 0, false
+	}
+	return 1, true
+}
+
+func compareBool(l, r bool) int {
+	if l == r {
+		return 0
+	}
+	if l == false {
+		return -1
+	}
+	return 1
+}
+
+// compareCandidates is the core of skyline pruning. It compares the two candidate paths on three dimensions:
+// (1): the set of columns that occurred in the access condition,
+// (2): whether or not it matches the physical property
+// (3): does it require a double scan.
+// If `x` is not worse than `y` at all factors,
+// and there exists one factor that `x` is better than `y`, then `x` is better than `y`.
+func compareCandidates(lhs, rhs *candidatePath) int {
+	setsResult, comparable := compareColumnSet(lhs.columnSet, rhs.columnSet)
+	if !comparable {
+		return 0
+	}
+	scanResult := compareBool(lhs.isSingleScan, rhs.isSingleScan)
+	matchResult := compareBool(lhs.isMatchProp, rhs.isMatchProp)
+	sum := setsResult + scanResult + matchResult
+	if setsResult >= 0 && scanResult >= 0 && matchResult >= 0 && sum > 0 {
+		return 1
+	}
+	if setsResult <= 0 && scanResult <= 0 && matchResult <= 0 && sum < 0 {
+		return -1
+	}
+	return 0
+}
+
+func (ds *DataSource) getTableCandidate(path *accessPath, prop *property.PhysicalProperty) *candidatePath {
+	candidate := &candidatePath{path: path}
+	pkCol := ds.getPKIsHandleCol()
+	candidate.isMatchProp = len(prop.Items) == 1 && pkCol != nil && prop.Items[0].Col.Equal(nil, pkCol)
+	candidate.columnSet = expression.ExtractColumnSet(path.accessConds)
+	candidate.isSingleScan = true
+	return candidate
+}
+
+func (ds *DataSource) getIndexCandidate(path *accessPath, prop *property.PhysicalProperty) *candidatePath {
+	candidate := &candidatePath{path: path}
```

Comment:
how about:
```go
candidate := &candidatePath{path: path}
candidate.columnSet = expression.ExtractColumnSet(path.accessConds)
candidate.isSingleScan = isCoveringIndex(ds.schema.Columns, path.index.Columns, ds.tableInfo.PKIsHandle)

all, _ := prop.AllSameOrder()
if prop.IsEmpty() {
	candidate.isMatchProp = true
	return candidate
}

if !all {
	// The order property contains multiple columns and not all of them
	// require the same asc/desc order, this kind of physical property
	// can not be matched in the current index implementation.
	candidate.isMatchProp = false
	return candidate
}

for i, col := range path.index.Columns {
	// ..
}
```

Reply:
When the prop is empty, `false` keeps consistent with the origin code. And the benefit of `false` is that in https://github.com/pingcap/tidb/pull/9337/files#diff-72b67e16ace76453819b4a58057284d6L551, we do not need to add `keep order` for the index scan.

Reply:
OK, could you add this info to the comment?

Reply:
ok.